csrc/id_model/id_model.h

```diff
@@ -0,0 +1,136 @@
+
+  // IterDomains from the original fusion are only allowed to be used once in
+  // the IterDomain graph, id->uses() are not directly used as there's no bounds
+  // check that would prevent a use from being defined that's not part of the
+  // actual fusion definition.
+  //
+  // Note, any iter domains used during something like loop or concrete id
+  // resolution could actually have multiple Expr* uses, and uses on disjoint id
+  // sets should be used, not this.
+  //
+  // TODO: Refactor or remove?
+  Expr* idUse(IterDomain* id) const;
+  Expr* idDef(IterDomain* id) const;
+
+  // TODO: Seems a bit unfortunate that this isn't IterDomain local information.
+  const std::unordered_set<IterDomain*>& viewRfactorIds() const {
+    return view_rfactor_ids_;
+  }
+
+  // Update the LOOP ID disjoint sets with resolved computeWith
+  void updateComputeWith(TensorView* compute_with_tv);
+
+  std::string toString() const;
+
+  // TODO: Should this not be private?
+ protected:
+  // Sometimes fusion inputs or outputs are disconnected from expressions, in
+  // those cases we still may want to send in some additional tensor views from
+  // the Fusion that don't have expressions associated with them.
+  void build(
+      const std::vector<Expr*>& exprs,
+      const std::vector<TensorView*>& additional_tvs);
+
+  // ======= START Iteration domain build process in order called =======
+
+  // Fills id_uses_ and id_definitions_ for all IterDomains active in the
+  // fusion.
+  void buildIterDomainDefinitionsAndUses(
+      const std::vector<TensorView*>& all_tvs);
+
+  // Iterates over all IterDomains in id_definitions_ and calls initializeID on
+  // a new IdGraph and returns it.
+  ValGraph initializeIdGraph(bool propagate_through_exprs = true);
+
+  // Fills disjoint_ids_[IdMappingMode::EXACT] for relationships between inputs
+  // and first output of expr
+  void buildExactGraph(const std::vector<Expr*>& exprs);
+
+  // Keeps a disjoint set entry for all IterDomain for all mapping mode types.
+  //
+  // Using an array here might be nice, but it seems hard to use an enum as an
+  // array key
+  // https://stackoverflow.com/questions/2102582/how-can-i-count-the-items-in-an-enum
+  std::unordered_map<IdMappingMode, ValGraph> id_graphs_;
+
+  // If multiple transformations occur IterDomains could have multiple uses,
+  // however only one should be active in the given Fusion. When we resolve loop
+  // promotions during lowering, we can generate new iter domains from existing
+  // ones, so there can be multiple uses generated. Tracks all the active iter
+  // domain uses.
+  std::unordered_map<IterDomain*, VectorOfUniqueEntries<Expr*>> id_uses_;
+
+  // Make sure we don't blindly use definitions as we don't want to grab
+  // transformations before a tensor view's root domain. There can be
+  // multiple definitions due to replays.
+  std::unordered_map<IterDomain*, VectorOfUniqueEntries<Expr*>> id_definitions_;
```

Comment:
Should we keep track of definitions and uses in `ValGraph`? The corresponding `idUse` and `idDef` should also be moved to `ValGraph`?

Reply:
Thanks for the feedback, but I am not sure. Part of the reason why it's currently designed this way would be that `IdModel` owns the whole ID information, and each graph is just a shallow representation of particular groupings and dependencies, and in that sense I feel it's more natural for `IdModel` to own the underlying use and def information.

Either way, I'd keep it as is and revisit as necessary.

Reply:
Then is it possible to make `initializeVal` not taking a list of definition and uses? This design of `initializeVal` makes `ValGraph` very difficult to use because it requires manual construction of list of definitions and uses. Can `initializeVal` just use `val->definition()` and `val->uses()`?

Reply:
I am actually experimenting with `ValGraph` by building a `ValGraph` for `PrecomputedValues`, so that equivalent vals are not evaluated repeatedly, therefore making runtime faster by spending more time at compile time. But I find't that I can not even build a `ValGraph` because of this `initializeVal`.

Reply:
> Can initializeVal just use val->definition() and val->uses()?

We avoided that because we may want to be able to have a root domain that has a definition. We would want to ignore the definition for this analysis.

Also, definitions are no longer limited to be unique.

Reply:
I think what we could do would be make the `defs` and `uses` parameters optional, and if not provided, `definition()` and `uses()` would be used. Would that enough for `PrecomputedValues`?

Reply:
> Also, definitions are no longer limited to be unique.

Why so? I am pretty sure there is a good reason. But I don't get why.



> I think what we could do would be make the `defs` and `uses` parameters optional, and if not provided, `definition()` and `uses()` would be used. Would that enough for `PrecomputedValues`?

I think it will.

Reply:
> Why so? I am pretty sure there is a good reason. But I don't get why.

That's actually a good question. I remember @csarofeen enabled that for domain replays and having `id_definitions_` as a vector is because that's a possibility. But I'm now not sure if it's actually still the case. I'll check that while working on the remaining pieces of the original PR (#32)

> I think it will.

OK, let me change it accordingly

Reply:
Added: [4d6811f](https://github.com/NVIDIA/Fuser/pull/1168/commits/4d6811f20fc50a66fda30b2bc13cb5ac1571f7e5)